File: implementations/react/src/patterns/structural/composite/object-composite.ts
import type { JsonComponent } from "./component";
import { buildNode } from "./helpers/node-builder";
import type { Json } from "./types/json";

export class ObjectNode implements JsonComponent {
  private kids: JsonComponent[];
  public keyLabel: string;
  constructor(keyLabel: string, obj: { [k: string]: Json }) {
    this.keyLabel = keyLabel;
    this.kids = Object.entries(obj).map(([k, v]) => buildNode(k, v));
  }
  isLeaf() { return false; }
  getPreview() { return `{… ${this.kids.length} }`; }
  getChildren() { return this.kids; }
}

File: implementations/flutter/design_patterns_flutter/lib/patterns/structural/composite/object_composite.dart
import 'component.dart';

class ObjectNode implements JsonComponent {
  @override
  final String keyLabel;
  final List<JsonComponent> _kids;

  ObjectNode(this.keyLabel, Map<String, dynamic> obj, JsonComponent Function(String, dynamic) buildNode)
      : _kids = obj.entries.map((e) => buildNode(e.key, e.value)).toList();

  @override
  bool isLeaf() => false;

  @override
  String getPreview() => '{… ${_kids.length} }';

  @override
  List<JsonComponent> getChildren() => _kids;
}
